### ISSUE TITLE
Internal server error for code-push release

### ISSUE THREAD

**Issue by nibhatish** [Microsoft/cordova-plugin-code-push#73]
I am testing code-push for new ionic app. I followed the guide and everything is setup correctly. 

1. I ran following command and got the internal server error. 
 code-push release codepushApp "platforms/android/assets/www" "0.0.2" --mandatory
Upload progress:[==================================================] 100% 0.0s
[Error]  Internal Server Error

I have tried this 4-5 times at least and every time, it's the same issue. It uploads zip file and once upload is done, Internal Server error is shown. In case, it makes any difference, I am trying to upload it from India. 

2. In case, I try to release single file or only templates folder, then it uploads successfully. 
code-push release codepushApp "platforms/android/assets/www/templates/tab-dash.html" "0.0.2" --mandatory
Upload progress:[==================================================] 100% 0.0s
Successfully released an update containing the "platforms/android/assets/www/templates/tab-dash.html" file to the "Staging" deployment of the "codepushApp" app.

Now, problem is when mobile app tries to update to new version, it gives error. If, it is single file, it gives error - Cannot unzip file.
If, it is templates folder - It gives error. - Start page of application not found.

Can we have some more details returned, instead of just Internal Server error.

**Comment by lostintangent** [comment]
The CodePush plugin expects you to release the www folder, so releasing anything else is expected to result in an error (e.g. the update isn't a zip, the index.html file can't be found).

That said, you definitely shouldn't be getting a server error when trying to release the www folder, so something must be up. You being in India should have an impact on this.

Let us investigate this and get back to you ASAP. Apologies for the inconvenience!

**Comment by geof90** [comment]
Hi @nibhatish, which version of code-push-cli are you using?

**Comment by nibhatish** [comment]
@geof90  My version is v1.7.0-beta.
@lostintangent  Yes. I expected those two attempts to not work. I just wanted to give more information regarding whatever I tried, in case it is useful.

**Comment by nibhatish** [comment]
@geof90 Today, I am able to upload without any issues. I am not getting internal server error. I will let you know, if I receive this error again and will run it with node-debug.

**Comment by geof90** [comment]
@nibhatish I see, great that it worked out. Feel free to ping us if you run into any more issues.

**Comment by nibhatish** [comment]
I am again getting same internal server error. I have attached screenshots of node-debug and final http response from Codepush server.
[CodePushResponse.docx](https://github.com/Microsoft/cordova-plugin-code-push/files/171884/CodePushResponse.docx)
. -

**Comment by nibhatish** [comment]
I again encountered same internal server error 5-6 times. I have attached screenshots of node-debug and final http response from Codepush server. 
[CodePushResponse.docx](https://github.com/Microsoft/cordova-plugin-code-push/files/173835/CodePushResponse.docx)

**Comment by dtivel** [comment]
Hi, Nibhatish.

Can you please try again?  I’m hoping that we’re capturing the right diagnostic information on our end to isolate the root cause.

Thanks,

Damon

**Comment by nibhatish** [comment]
I tried it couple of times. There was no error first time. Second time - it gave internal server error. 
Approx. GMT Time :- 12:09 PM

**Comment by dtivel** [comment]
Hi, Nibhatish.

I have looked at logs of your codepushApp release activity, and I do not see any evidence of related HTTP 500 ("internal server error") responses.  I see either an HTTP 201 or an HTTP 409.  An HTTP 201 is returned when a release was successfully created.  An HTTP 409 is returned when there is a conflict that fails the release; attempting to release the same update payload that was just previously successfully released would generate an HTTP 409 with the following message:

	The uploaded package is identical to the contents of the specified deployment's current release.

Let's suppose that I build my update and release it to my CodePush "Staging" environment twice.  The first release should succeed with an HTTP 201, and the second release should fail with an HTTP 409, because the update's contents did not change between releases.  This is what your release activity looks like (especially on March 11th and 14th).  I see a 201 followed by a 409.  Is it possible that in your release attempts the failed releases were using update payloads that were identical to a previous successful release?  This would be useful to know:  if you are seeing an HTTP 500 response whenever you would normally see an HTTP 409 response (because the update payload has not changed).  If this pattern were true, I would wonder if there is an intermediate agent which is converting an HTTP 409 response into a HTTP 500.  Are you aware of any such possibility?

Can you try a couple of tests?

First, extract the contents of the attached [test.zip](https://github.com/Microsoft/cordova-plugin-code-push/files/182457/test.zip) file into a directory, open a command prompt in that directory and run:

    node test.js

...and post the result.  This test will try receiving a handful of status codes from https://httpbin.org.

Second, can you try releasing to CodePush again, mindful of the expected pattern for successful and failed releases, to see if the pattern fits (excepting the fact that you're seeing HTTP 500's instead of 409's).

Thanks,

Damon

**Comment by nibhatish** [comment]
Hello @dtivel , I had received the http 409 response when it successfully uploaded the release just after my unsuccessful attempt. I receive this message, if I try to upload the same version. 
The uploaded package is identical to the contents of the specified deployment's current release.

So, I don't think it's converting 409 to 501. I have received this error, even when release is absolutely fine and uploads successfully next time. 

One thing, which I have observed is , generally when upload takes more time ( ~ 2 minutes) due to slow connection, I receive Internal server error. Is there any Code push release process which  depends on some action to be completed in specific time ?

**Comment by nibhatish** [comment]
@dtivel PFB response for node test.js. 

Method  URL                             Expected  Actual  Result
GET     https://httpbin.org/status/200  200       200     OK
GET     https://httpbin.org/status/201  201       201     OK
GET     https://httpbin.org/status/204  204       204     OK
GET     https://httpbin.org/status/304  304       304     OK
GET     https://httpbin.org/status/400  400       400     OK
GET     https://httpbin.org/status/401  401       401     OK
GET     https://httpbin.org/status/403  403       403     OK
GET     https://httpbin.org/status/404  404       404     OK
GET     https://httpbin.org/status/409  409       409     OK
GET     https://httpbin.org/status/500  500       500     OK
POST    https://httpbin.org/status/200  200       200     OK
POST    https://httpbin.org/status/201  201       201     OK
POST    https://httpbin.org/status/204  204       204     OK
POST    https://httpbin.org/status/304  304       304     OK
POST    https://httpbin.org/status/400  400       400     OK
POST    https://httpbin.org/status/401  401       401     OK
POST    https://httpbin.org/status/403  403       403     OK
POST    https://httpbin.org/status/404  404       404     OK
POST    https://httpbin.org/status/409  409       409     OK
POST    https://httpbin.org/status/500  500       500     OK

**Comment by dtivel** [comment]
Thanks, Nibhatish.

A lot of stuff happens in during a release; but no, there aren't any time-based expectations in the release path.

That's interesting that you say that the failures occur when uploads take a long time (upwards of 2 minutes).  Another issue similar to this was just [reported](https://github.com/Microsoft/code-push/issues/175) with ~30 seconds between request start and failure.

Are you using the CodePush CLI behind a proxy server, like say at a work?

**Comment by nibhatish** [comment]
I am using it from my home.

**Comment by mccraigmccraig** [comment]
i am also seeing something that looks a lot like this error - it only seems to happen when i upload dev builds, which are quite large (21Mb). if i upload the much smaller production build (8.5Mb) then it seems to work ok

**Comment by dtivel** [comment]
Apologies for the inconvenience.

We were hitting a Node.js default connection timeout of 2 minutes.  Uploading a large file over a slow network connection could hit this timeout.

I have increased the timeout.  Also, the server should now return an HTTP 408 (request timeout) instead of a 500 (internal server error) in case of a timeout.

Can you please try again and let me know the result?

Thanks,

Damon

**Comment by nibhatish** [comment]
@dtivel  I can confirm that, now I am getting Request Timeout error for .zip file of size around 21 MB and not internal server error. I would suggest to increase timeout to at least 600 seconds as 21 MB is quite acceptable size for mobile app and for slow connections it might take up to 10 minutes to upload app.

**Comment by dtivel** [comment]
The timeout was 600 seconds.  Your request timed out because it took longer than 600 seconds.  We increased the timeout again.  Can you please let me know if it works for you now?
